book/src/design/circuit/gadgets/sinsemilla/commit-ivk.md

```diff
@@ -0,0 +1,229 @@
+enforce that these are indeed canonically-encoded field elements, i.e.
+
+$$
+\begin{align}
+\ItoLEBSP{\BaseLength{Orchard}}(\AuthSignPublic) &< q_\mathbb{P} \\
+\ItoLEBSP{\BaseLength{Orchard}}(\NullifierKey) &< q_\mathbb{P} \\
+\end{align}
+$$
+
+The Pallas base field modulus has the form $q_\mathbb{P} = 2^{254} + t_\mathbb{P}$, where
+$$t_\mathbb{P} = \mathtt{0x224698fc094cf91b992d30ed00000001}$$
+is 126 bits. We therefore know that if the top bit is not set, then the remaining bits
+will always comprise a canonical encoding of a field element. Thus the canonicity checks
+below are enforced if and only if $b_1 = 1$ (for $\AuthSignPublic$) or $d_1 = 1$ (for
+$\NullifierKey$).
+
+> In the constraints below we use a base-$2^{10}$ variant of the method used in libsnark
+> (originally from [[SVPBABW2012](https://eprint.iacr.org/2012/598.pdf), Appendix C.1]) for
+> range constraints $0 \leq x < t$:
+>
+> - Let $t'$ be the smallest power of $2^{10}$ greater than $t$.
+> - Enforce $0 \leq x < t'$.
+> - Let $x' = x + t' - t$.
+> - Enforce $0 \leq x' < t'$.
+
+### $\AuthSignPublic$ with $b_1 = 1 \implies \AuthSignPublic \geq 2^{254}$
+
+In these cases, we check that $\textsf{ak}_{0..=253} < t_\mathbb{P} < 2^{126}$:
+
+1. $b_1 = 1 \implies b_0 = 0.$
+
+   Since $b_1 = 1 \implies \AuthSignPublic_{0..=253} < 2^{126},$ we know that
+   $\AuthSignPublic_{126..=253} = 0,$ and in particular
+   $$b_0 := \AuthSignPublic_{250..=253} = 0.$$
+
+2. $b_1 = 1 \implies 0 \leq a < 2^{126}.$
+
+   To check that $a < 2^{126}$, we use two constraints:
+
+    a) $0 \leq a < 2^{130}$. This is expressed in the custom gate as
+       $$b_1 \cdot z_{a,13} = 0,$$
+       where $z_{a,13}$ is the index-13 running sum output by $\SinsemillaHash(a).$
+
+    b) $0 \leq a + 2^{130} - t_\mathbb{P} < 2^{130}$. To check this, we decompose
+       $a' = a + 2^{130} - t_\mathbb{P}$ into thirteen 10-bit words (little-endian) using
+       a running sum $z_{a'}$, looking up each word in a $10$-bit lookup table. We then
+       enforce in the custom gate that
+       $$b_1 \cdot z_{a',13} = 0.$$
+
+$$
+\begin{array}{|c|l|}
+\hline
+\text{Degree} & \text{Constraint} \\\hline
+3 & q_\CommitIvk \cdot b_1 \cdot b_0 = 0 \\\hline
+3 & q_\CommitIvk \cdot b_1 \cdot z_{a,13} = 0 \\\hline
+2 & q_\CommitIvk \cdot (a + 2^{130} - t_\mathbb{P} - a') = 0 \\\hline
+3 & q_\CommitIvk \cdot b_1 \cdot z_{a',13} = 0 \\\hline
+\end{array}
+$$
+
+### $\NullifierKey$ with $d_1 = 1 \implies \NullifierKey \geq 2^{254}$
+
+In these cases, we check that $\textsf{nk}_{0..=253} < t_\mathbb{P} < 2^{126}$:
+
+1. $d_1 = 1 \implies d_0 = 0.$
+
+   Since $d_1 = 1 \implies \NullifierKey_{0..=253} < 2^{126},$ we know that $\NullifierKey_{126..=253} = 0,$ and in particular $$d_0 := \NullifierKey_{245..=253} = 0.$$
+
+2. $d_1 = 1 \implies 0 \leq b_2 + 2^5 \cdot c < 2^{126}.$
+
+   To check that $0 \leq b_2 + 2^5 \cdot c < 2^{126}$, we use two constraints:
```

Comment:
```suggestion
In these cases, we check that $\textsf{nk}_{0..=253} < t_\mathbb{P}$:

1. $d_1 = 1 \implies d_0 = 0.$

   Since $d_1 = 1 \implies \NullifierKey_{0..=253} < t_\mathbb{P} < 2^{126},$ we know that $\NullifierKey_{126..=253} = 0,$ and in particular $$d_0 := \NullifierKey_{245..=253} = 0.$$

2. $d_1 = 1 \implies 0 \leq b_2 + 2^5 \cdot c < t_\mathbb{P}.$

   To check that $0 \leq b_2 + 2^5 \cdot c < t_\mathbb{P}$, we use two constraints:
```